particle-iot/device-os
# Clear module slots in DCT when preparing for an OTA update

## Note: This PR is targeting `feature/refactor-system-info-generation` (https://github.com/particle-iot/device-os/pull/2347).

### Problem

When system resets are disabled via `System.disableReset()`, it is possible to flash the device OTA multiple times without it resetting itself to apply the first received update. Module slots that are stored in the DCT and trigger the bootloader to copy the module's data from the OTA section to its destination section in the internal flash are not cleared in this case, even if the received modules occupy the same area in the OTA section, which may result in a corrupted firmware update.

### Solution

Clear all non-factory module slots in the DCT at the beginning of an OTA update.

### Steps to Test

1. Flash the example application to a Gen 2 or Gen 3 device. Let the device connect to the cloud.
2. Flash the main system part module that is already running on the device OTA. The device will not reset after the update and stay connected to the cloud.
3. Flash some firmware module that has unmet dependencies to the device OTA. One way to get such a module is to bump the `BOOTLOADER_DEPENDENCY` version of the main system part module that is running on the device.
4. Reset the device via the reset button and verify that the bootloader doesn't try to apply any updates. Running this test without the fix in place would result in the bootloader copying the module with unmet dependencies to its destination section in the internal flash.

### Example App

```c
#include "application.h"

SYSTEM_MODE(SEMI_AUTOMATIC)
SYSTEM_THREAD(ENABLED)

void setup() {
    System.disableReset();
    Particle.connect();
}

void loop() {
}
```

### References

- [ch84957]
